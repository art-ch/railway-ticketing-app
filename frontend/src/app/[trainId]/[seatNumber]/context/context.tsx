'use client';

import {
  createContext,
  useCallback,
  useContext,
  useEffect,
  useState
} from 'react';
import { useTrainPageContext } from '../../context';
import { Booking as BookingType } from 'railway-ticketing-app-sdk';
import { axiosInstance } from '@/api/axios/axios';

/** Using `undefined` as initial state to indicate "not yet fetched" */
export type BookingData = BookingType | null | undefined;

export type BookingPageContextProps = {
  bookingData: BookingData;
  isBookingLoading: boolean;
  updateBooking: (
    seatNumber: number,
    bookingId?: string,
    newBooking?: BookingType | null
  ) => void;
};

export const BookingPageContext = createContext<BookingPageContextProps | null>(
  null
);

export type BookingPageProviderProps = {
  children: React.ReactNode;
  seatNumber: string;
};

export const BookingPageProvider = ({
  children,
  seatNumber
}: BookingPageProviderProps) => {
  const { trainData, setTrainData } = useTrainPageContext();

  const [bookingData, setBookingData] = useState<BookingData>(undefined);
  const [isBookingLoading, setIsBookingLoading] = useState(false);

  const fetchBooking = useCallback(
    async (bookingId: string) => {
      setIsBookingLoading(true);

      try {
        const response = await axiosInstance.get(`/bookings/${bookingId}`);
        setBookingData(response.data.data);
      } catch (error) {
        console.error(`Error fetching booking for seat ${seatNumber}:`, error);
        setBookingData(null);
      } finally {
        setIsBookingLoading(false);
      }
    },
    [seatNumber]
  );

  useEffect(() => {
    if (!trainData) return;

    const seatData = trainData.seats.find(
      (seat) => seat.seatNumber.toString() === seatNumber
    );

    if (!seatData) {
      setBookingData(null);
      return;
    }

    if (seatData.bookingId) {
      fetchBooking(seatData.bookingId);
    } else {
      // No booking for this seat
      setBookingData(null);
    }
  }, [fetchBooking, trainData, seatNumber]);

  const updateBooking = (
    seatNumber: number,
    bookingId?: string,
    newBooking?: BookingType | null
  ) => {
    // Update the train data
    updateTrainSeatData(seatNumber, bookingId);

    // Update the booking data based on the scenario
    handleBookingDataUpdate(bookingId, newBooking);
  };

  const updateTrainSeatData = (seatNumber: number, bookingId?: string) => {
    setTrainData((prevData) => {
      if (!prevData) return null;

      const updatedSeats = prevData.seats.map((seat) =>
        seat.seatNumber === seatNumber ? { ...seat, bookingId } : seat
      );

      return {
        ...prevData,
        seats: updatedSeats
      };
    });
  };

  const handleBookingDataUpdate = (
    bookingId?: string,
    newBooking?: BookingType | null
  ) => {
    // Case 1: Explicit booking data provided
    if (newBooking !== undefined) {
      setBookingData(newBooking);
      return;
    }

    // Case 2: New booking ID provided but no data
    if (bookingId) {
      fetchBooking(bookingId);
      return;
    }

    // Case 3: No booking ID (seat is now unbooked)
    setBookingData(null);
  };

  return (
    <BookingPageContext.Provider
      value={{
        bookingData,
        isBookingLoading,
        updateBooking
      }}
    >
      {children}
    </BookingPageContext.Provider>
  );
};

export const useBookingPageContext = () => {
  const context = useContext(BookingPageContext);

  if (!context) {
    throw new Error(
      'useBookingPageContext must be used within a TrainPageProvider'
    );
  }

  return context;
};
