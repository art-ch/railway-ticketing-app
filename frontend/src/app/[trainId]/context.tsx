'use client';
import { createContext, useContext, useState } from 'react';

import { SetState } from '@/types/types';
import { Train } from 'railway-ticketing-app-sdk';

export type TrainPageData = Train | null;

export type TrainPageContextProps = {
  trainData: TrainPageData;
  setTrainData: SetState<TrainPageData>;
};

export const TrainPageContext = createContext<TrainPageContextProps | null>(
  null
);

export type TrainPageProviderProps = {
  children: React.ReactNode;
  pageData: TrainPageData;
};

export const TrainPageProvider = ({
  children,
  pageData
}: TrainPageProviderProps) => {
  const [trainData, setTrainData] = useState<TrainPageData>(pageData);

  return (
    <TrainPageContext.Provider
      value={{
        trainData,
        setTrainData
      }}
    >
      {children}
    </TrainPageContext.Provider>
  );
};
export const useTrainPageContext = () => {
  const context = useContext(TrainPageContext);

  if (!context) {
    throw new Error(
      'useTrainPageContext must be used within a TrainPageProvider'
    );
  }

  return context;
};
