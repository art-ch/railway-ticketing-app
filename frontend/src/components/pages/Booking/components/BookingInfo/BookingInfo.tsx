import React from 'react';

import { Booking as BookingType } from '@/models';
import { getFormattedDateTime } from '@/utils/formatters';

export type BookingInfoProps = { bookingInfo: BookingType };

export const BookingInfo = ({ bookingInfo }: BookingInfoProps) => {
  const { seatNumber, status, bookingTime } = bookingInfo;

  return (
    <div>
      <p className="font-bold pb-1">Booking Info</p>

      <table className="w-full">
        <thead>
          <tr className="text-sm font-semibold">
            <th className="text-left">Seat</th>
            <th className="text-left">Status</th>
            <th className="text-left">Booked at</th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>{seatNumber}</td>
            <td>{status}</td>
            <td>{getFormattedDateTime(new Date(bookingTime))}</td>
          </tr>
        </tbody>
      </table>
    </div>
  );
};
