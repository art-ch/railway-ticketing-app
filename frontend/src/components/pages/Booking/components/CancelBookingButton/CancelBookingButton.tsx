import React, { useState } from 'react';

import { Booking as BookingType } from 'railway-ticketing-app-sdk';
import { Button } from '@/components/ui/button';
import { axiosInstance } from '@/api/axios/axios';
import { LoaderCircle, TicketX as TicketXIcon } from 'lucide-react';

export type CancelBookingButtonProps = {
  bookingInfo: BookingType;
  isBookingLoading: boolean;
  onSuccess: (receivedData: BookingType) => void;
  onError: () => void;
};

export const CancelBookingButton = ({
  bookingInfo,
  isBookingLoading,
  onSuccess,
  onError
}: CancelBookingButtonProps) => {
  const [isCancellationInProgress, setIsCancellationInProgress] =
    useState(false);

  const handleSubmit = async () => {
    setIsCancellationInProgress(true);

    try {
      const response = await axiosInstance.patch(
        `/bookings/${bookingInfo.bookingId}`,
        {
          status: 'CANCELLED'
        }
      );

      const receivedData = response.data.data || null;

      onSuccess(receivedData);
    } catch (error) {
      console.error(error);

      onError();
    } finally {
      setIsCancellationInProgress(false);
    }
  };

  const isBookingBeingProcessed = isCancellationInProgress || isBookingLoading;

  return (
    <Button
      className="size-sm self-end mt-4 bg-red-600 hover:bg-red-500 cursor-pointer disabled:cursor-not-allowed"
      onClick={() => handleSubmit()}
      disabled={isBookingBeingProcessed}
    >
      {isBookingBeingProcessed ? (
        <LoaderCircle className="animate-spin" />
      ) : (
        <TicketXIcon />
      )}
      Cancel Booking
    </Button>
  );
};
