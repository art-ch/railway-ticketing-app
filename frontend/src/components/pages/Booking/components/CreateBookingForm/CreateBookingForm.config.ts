import { PassengerDetails } from '@/models/passengerDetails.model';

export const CREATE_BOOKING_FORM_FIELD_CONFIG: {
  name: keyof PassengerDetails;
  label: string;
  type: React.HTMLInputTypeAttribute;
  placeholder: string;
}[] = [
  { name: 'name', label: 'Name', type: 'text', placeholder: 'Type in a name' },
  {
    name: 'email',
    label: 'Email',
    type: 'email',
    placeholder: 'Type in an email'
  },
  {
    name: 'phone',
    label: 'Phone number (optional)',
    type: 'tel',
    placeholder: 'Type in phone number'
  }
];
