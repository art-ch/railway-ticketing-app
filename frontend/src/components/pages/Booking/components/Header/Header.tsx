import React from 'react';

export type HeaderProps = { bookingId?: string };

export const Header = ({ bookingId }: HeaderProps) => (
  <div>
    <p className="absolute left-4 top-4 border-2 border-black px-1">22</p>
    <div className="flex gap-2 items-baseline justify-center font-bold pb-2">
      <p className="text-sm">art-ch railway</p>
      <p className="text-lg border-l border-r border-black px-2">
        Travel Document
      </p>
      {!bookingId && <p className="text-sm px-2 text-gray-300">booking id</p>}
      {bookingId && (
        <p className="text-sm text-nowrap w-33">
          {/** shortening ID for demo purposes */}
          ID #{bookingId.replace(/(.*?)-.*/, '$1')}
        </p>
      )}
    </div>
  </div>
);
