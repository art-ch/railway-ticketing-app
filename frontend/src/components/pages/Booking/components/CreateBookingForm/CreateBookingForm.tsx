'use client';

import React, { useState } from 'react';

import { Ticket as TicketIcon, LoaderCircle } from 'lucide-react';

import { Button } from '@/components/ui/button';
import {
  FormField as ShadCnFormField,
  Form as ShadCnForm,
  FormItem,
  FormLabel,
  FormControl,
  FormMessage
} from '@/components/ui/form';
import { Input } from '@/components/ui/input';
import { zodResolver } from '@hookform/resolvers/zod';
import { useForm } from 'react-hook-form';
import {
  PassengerDetails,
  PassengerDetailsSchema
} from 'railway-ticketing-app-sdk';
import { useTrainPageContext } from '@/app/[trainId]/context';
import { axiosInstance } from '@/api/axios/axios';
import { Booking as BookingType } from 'railway-ticketing-app-sdk';
import { CREATE_BOOKING_FORM_FIELD_CONFIG } from './CreateBookingForm.config';

export type CreateBookingFormProps = {
  seatNumber: string;
  isBookingLoading: boolean;
  onSuccess: (receivedData: BookingType) => void;
  onError: () => void;
};

export const CreateBookingForm = ({
  seatNumber,
  isBookingLoading,
  onSuccess,
  onError
}: CreateBookingFormProps) => {
  const [isCreationInProgress, setIsCreationInProgress] = useState(false);

  const { trainData: trainInfo } = useTrainPageContext();

  const form = useForm({
    resolver: zodResolver(PassengerDetailsSchema),
    defaultValues: {
      name: '',
      email: '',
      phone: ''
    }
  });

  if (!trainInfo) {
    return <div>No such train found</div>;
  }

  const handleFormSubmit = async (passengerDetails: PassengerDetails) => {
    setIsCreationInProgress(true);

    try {
      const response = await axiosInstance.post('bookings', {
        trainId: trainInfo.trainId,
        seatNumber: Number(seatNumber),
        passengerDetails
      });

      const receivedData = response.data.data || null;

      onSuccess(receivedData);
    } catch (error) {
      console.error(error);

      onError();
    } finally {
      setIsCreationInProgress(false);
    }
  };

  const isBookingBeingProcessed = isCreationInProgress || isBookingLoading;

  return (
    <ShadCnForm {...form}>
      <form onSubmit={form.handleSubmit(handleFormSubmit)}>
        <p className="font-bold pb-1">Passenger Info</p>
        <div className="flex flex-col items-end gap-4">
          <div className="flex gap-2">
            {CREATE_BOOKING_FORM_FIELD_CONFIG.map((fieldConfig) => (
              <ShadCnFormField
                key={fieldConfig.name}
                control={form.control}
                name={fieldConfig.name}
                render={({ field }) => (
                  <FormItem>
                    <FormLabel>{fieldConfig.label}</FormLabel>
                    <FormControl>
                      <Input
                        className="bg-white h-6 p-2"
                        type={fieldConfig.type}
                        placeholder={fieldConfig.placeholder}
                        {...field}
                      />
                    </FormControl>
                    <FormMessage />
                  </FormItem>
                )}
              />
            ))}
          </div>
          <Button
            className="flex gap-2 cursor-pointer disabled:cursor-not-allowed bg-blue-800 hover:bg-blue-600"
            type="submit"
            size="sm"
            disabled={isBookingBeingProcessed}
          >
            {isBookingBeingProcessed ? (
              <LoaderCircle className="animate-spin" />
            ) : (
              <TicketIcon />
            )}
            Book a seat
          </Button>
        </div>
      </form>
    </ShadCnForm>
  );
};
