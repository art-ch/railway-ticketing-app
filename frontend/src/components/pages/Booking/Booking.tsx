'use client';

import React from 'react';
import { CreateBookingForm } from './components/CreateBookingForm/CreateBookingForm';
import { Header } from './components/Header/Header';
import { TrainInfo } from './components/TrainInfo/TrainInfo';
import { Booking as BookingType } from 'railway-ticketing-app-sdk';
import { BookingInfo } from './components/BookingInfo/BookingInfo';
import { PassengerInfo } from './components/PassengerInfo/PassengerInfo';
import { CancelBookingButton } from './components/CancelBookingButton/CancelBookingButton';
import { toast } from 'sonner';
import { useBookingPageContext } from '@/app/[trainId]/[seatNumber]/context/context';
import { BookingDataSkeleton } from './components/BookingDataSkeleton/BookingDataSkeleton';
import { DemoWrapper } from '@/components/DemoWrapper/DemoWrapper';

export type BookingProps = {
  seatNumber: string;
};

export const Booking = ({ seatNumber }: BookingProps) => {
  const { bookingData, isBookingLoading, updateBooking } =
    useBookingPageContext();

  // Success callbacks
  const handleBookingSuccess = (newBookingData: BookingType) => {
    updateBooking(Number(seatNumber), newBookingData.bookingId, newBookingData);

    toast.success('Booking successfully created!');
  };

  const handleCancellationSuccess = () => {
    updateBooking(Number(seatNumber), undefined, null);

    toast.success('Booking successfully cancelled!');
  };

  // Error callbacks
  const handleBookingError = () => {
    toast.error('Failed to create booking, please try again later');
  };

  const handleCancellationError = () => {
    toast.error('Failed to cancel booking, please try again later');
  };

  const renderBookingInfo = () => {
    switch (true) {
      case isBookingLoading || bookingData === undefined:
        return <BookingDataSkeleton />;

      case bookingData === null:
        return (
          <CreateBookingForm
            seatNumber={seatNumber}
            isBookingLoading={isBookingLoading}
            onSuccess={handleBookingSuccess}
            onError={handleBookingError}
          />
        );

      case !!bookingData:
        return (
          <div className="flex flex-col">
            <BookingInfo bookingInfo={bookingData} />
            <PassengerInfo passengerInfo={bookingData.passengerDetails} />
            <CancelBookingButton
              bookingInfo={bookingData}
              isBookingLoading={isBookingLoading}
              onSuccess={handleCancellationSuccess}
              onError={handleCancellationError}
            />
          </div>
        );

      default:
        return null;
    }
  };

  return (
    <DemoWrapper>
      <main className="h-[calc(100vh-66px)] flex flex-col items-center justify-center">
        <div className="relative bg-amber-50 shadow p-4">
          <Header bookingId={bookingData?.bookingId} />
          <TrainInfo />
          {renderBookingInfo()}
        </div>
      </main>
    </DemoWrapper>
  );
};
