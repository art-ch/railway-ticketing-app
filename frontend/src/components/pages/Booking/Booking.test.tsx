import {
  BreakpointState,
  useBreakpoint
} from '@/hooks/useBreakpoint/useBreakpoint';
import { render, screen } from '@testing-library/react';
import { Booking } from './Booking';
import { TrainPageData, TrainPageProvider } from '@/app/[trainId]/context';
import {
  BookingPageProvider,
  useBookingPageContext
} from '@/app/[trainId]/[seatNumber]/context';
import { CreateBookingFormProps } from './components/CreateBookingForm/CreateBookingForm';
import { Booking as BookingType } from 'railway-ticketing-app-sdk';
import { BookingInfoProps } from './components/BookingInfo/BookingInfo';
import { CancelBookingButtonProps } from './components/CancelBookingButton/CancelBookingButton';
import { HeaderProps } from './components/Header/Header';
import userEvent from '@testing-library/user-event';
import { MouseEventHandler } from 'react';
import { toast } from 'sonner';

jest.mock('./components/BookingDataSkeleton/BookingDataSkeleton', () => ({
  BookingDataSkeleton: () => (
    <div data-testid="booking-skeleton">Loading...</div>
  )
}));

jest.mock('sonner', () => ({
  toast: {
    success: jest.fn(),
    error: jest.fn()
  }
}));

jest.mock('./components/CreateBookingForm/CreateBookingForm', () => ({
  CreateBookingForm: ({
    onSuccess,
    onError
  }: Pick<CreateBookingFormProps, 'onSuccess' | 'onError'>) => (
    <div data-testid="create-booking-form">
      <button
        onClick={() =>
          onSuccess({ bookingId: 'new-booking-123' } as BookingType)
        }
      >
        Create Booking
      </button>
      <button onClick={onError}>Trigger Error</button>
    </div>
  )
}));

jest.mock('./components/BookingInfo/BookingInfo', () => ({
  BookingInfo: ({ bookingInfo }: BookingInfoProps) => (
    <div data-testid="booking-info">{bookingInfo.bookingId}</div>
  )
}));

jest.mock('./components/PassengerInfo/PassengerInfo', () => ({
  PassengerInfo: () => <div data-testid="passenger-info">Passenger Details</div>
}));

jest.mock('./components/CancelBookingButton/CancelBookingButton', () => ({
  CancelBookingButton: ({
    onSuccess,
    onError
  }: Pick<CancelBookingButtonProps, 'onError' | 'onSuccess'>) => (
    <div data-testid="cancel-booking">
      <button
        onClick={onSuccess as unknown as MouseEventHandler<HTMLButtonElement>}
      >
        Cancel Booking
      </button>
      <button onClick={onError}>Trigger Cancel Error</button>
    </div>
  )
}));

jest.mock('./components/Header/Header', () => ({
  Header: ({ bookingId }: HeaderProps) => (
    <div data-testid="header">{bookingId || 'No Booking'}</div>
  )
}));

jest.mock('./components/TrainInfo/TrainInfo', () => ({
  TrainInfo: () => <div data-testid="train-info">Train Information</div>
}));

jest.mock('@/hooks/useBreakpoint/useBreakpoint', () => ({
  useBreakpoint: jest.fn()
}));

jest.mock('@/app/[trainId]/[seatNumber]/context', () => ({
  ...jest.requireActual('@/app/[trainId]/[seatNumber]/context'),
  useBookingPageContext: jest.fn().mockReturnValue({
    bookingData: {},
    isBookingLoading: false,
    updateBooking: jest.fn()
  })
}));

const useBookingPageContextMock = useBookingPageContext as jest.MockedFunction<
  typeof useBookingPageContext
>;
const useBreakpointMock = useBreakpoint as jest.MockedFunction<
  typeof useBreakpoint
>;

const renderComponent = (trainPageData: TrainPageData, seatNumber: string) =>
  render(
    <TrainPageProvider pageData={trainPageData}>
      <BookingPageProvider seatNumber={seatNumber}>
        <Booking seatNumber={seatNumber} />
      </BookingPageProvider>
    </TrainPageProvider>
  );

describe('Booking component', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    useBreakpointMock.mockReturnValue({ xl: true } as BreakpointState);
  });

  it('should show demo notice for small screen', () => {
    useBreakpointMock.mockReturnValue({ xl: false } as BreakpointState);

    const { container } = renderComponent(null, '1');

    expect(container).toMatchSnapshot();
  });

  it('should render loading state when booking data is loading', () => {
    useBookingPageContextMock.mockReturnValue({
      bookingData: undefined,
      isBookingLoading: true,
      updateBooking: jest.fn()
    });

    renderComponent(null, '1');

    expect(screen.getByTestId('booking-skeleton')).toBeInTheDocument();
    expect(screen.getByTestId('header')).toBeInTheDocument();
    expect(screen.getByTestId('train-info')).toBeInTheDocument();
  });

  it('should render create booking form when no booking exists', () => {
    useBookingPageContextMock.mockReturnValue({
      bookingData: null,
      isBookingLoading: false,
      updateBooking: jest.fn()
    });

    renderComponent(null, '1');

    expect(screen.getByTestId('create-booking-form')).toBeInTheDocument();
    expect(screen.getByTestId('header')).toBeInTheDocument();
    expect(screen.getByTestId('train-info')).toBeInTheDocument();
  });

  it('should render booking info when booking exists', () => {
    const mockBookingData = {
      bookingId: 'booking-123',
      passengerDetails: {
        name: 'John Doe',
        email: 'john@example.com'
      }
    } as BookingType;

    useBookingPageContextMock.mockReturnValue({
      bookingData: mockBookingData,
      isBookingLoading: false,
      updateBooking: jest.fn()
    });

    renderComponent(null, '1');

    expect(screen.getByTestId('booking-info')).toBeInTheDocument();
    expect(screen.getByTestId('passenger-info')).toBeInTheDocument();
    expect(screen.getByTestId('cancel-booking')).toBeInTheDocument();
    expect(screen.getByTestId('header')).toHaveTextContent('booking-123');
  });

  it('should handle booking creation success', async () => {
    const mockUpdateBooking = jest.fn();
    useBookingPageContextMock.mockReturnValue({
      bookingData: null,
      isBookingLoading: false,
      updateBooking: mockUpdateBooking
    });

    renderComponent(null, '2');

    const createButton = screen.getByText('Create Booking');
    await userEvent.click(createButton);

    expect(mockUpdateBooking).toHaveBeenCalledWith(2, 'new-booking-123', {
      bookingId: 'new-booking-123'
    });
    expect(toast.success).toHaveBeenCalledWith('Booking successfully created!');
  });

  it('should handle booking creation error', async () => {
    useBookingPageContextMock.mockReturnValue({
      bookingData: null,
      isBookingLoading: false,
      updateBooking: jest.fn()
    });

    renderComponent(null, '2');

    const errorButton = screen.getByText('Trigger Error');
    await userEvent.click(errorButton);

    expect(toast.error).toHaveBeenCalledWith(
      'Failed to create booking, please try again later'
    );
  });

  it('should handle booking cancellation success', async () => {
    const mockUpdateBooking = jest.fn();
    const mockBookingData = {
      bookingId: 'booking-123',
      passengerDetails: {
        name: 'John Doe',
        email: 'john@example.com'
      }
    } as BookingType;

    useBookingPageContextMock.mockReturnValue({
      bookingData: mockBookingData,
      isBookingLoading: false,
      updateBooking: mockUpdateBooking
    });

    renderComponent(null, '3');

    const cancelButton = screen.getByText('Cancel Booking');
    await userEvent.click(cancelButton);

    expect(mockUpdateBooking).toHaveBeenCalledWith(3, undefined, null);
    expect(toast.success).toHaveBeenCalledWith(
      'Booking successfully cancelled!'
    );
  });

  it('should handle booking cancellation error', async () => {
    const mockBookingData = {
      bookingId: 'booking-123',
      passengerDetails: {
        name: 'John Doe',
        email: 'john@example.com'
      }
    } as BookingType;

    useBookingPageContextMock.mockReturnValue({
      bookingData: mockBookingData,
      isBookingLoading: false,
      updateBooking: jest.fn()
    });

    renderComponent(null, '3');

    const errorButton = screen.getByText('Trigger Cancel Error');
    await userEvent.click(errorButton);

    expect(toast.error).toHaveBeenCalledWith(
      'Failed to cancel booking, please try again later'
    );
  });
});
