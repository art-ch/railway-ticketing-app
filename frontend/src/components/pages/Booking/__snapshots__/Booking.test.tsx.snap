// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Booking component should show demo notice for small screen 1`] = `
<div>
  <div>
    This demo is intended to be shown on desktop
  </div>
</div>
`;
